### PR TITLE
rbd: add a workaround to fix rbd snapshot scheduling

### PR DESCRIPTION
**Pull request by Madhu-1**
currently, we have a bug in the rbd mirror scheduling module. After doing failover and failback the scheduling is not getting updated and the mirroring snapshots are not getting created periodically as per the scheduling interval. This PR workaround this one by doing the below operations
    
* Create a dummy (csi-vol-dummy-ceph fsID) image per cluster and this image should be easily identified.
    
* During Promote operation on any image enables the mirroring on the dummy image. when we enable the mirroring on the dummy image the pool will get updated and the scheduling will be reconfigured.
    
* During Demote operation on any image disables the mirroring on the dummy image. the disabled need to be done to enable the mirroring again when we get the promote request to make the image as the primary
    
* When the DR is no more needed, this image needs to be  manually cleanup for now as we don't want to add a check
in the existing DeleteVolume code path for deleting dummy images as it impacts the performance of the DeleteVolume workflow.
 

Moved to add scheduling to the promote operation as scheduling need to be added when the image is promoted and this is the correct method of adding the scheduling to make the scheduling take place.

More details at  https://bugzilla.redhat.com/show_bug.cgi?id=2019161
Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

### CODE REVIEW

**Comment by Madhu-1** [comment]
Manual testing works for me. For now added DNM as its getting tested by @ShyamsundarR and @BenamarMk

File: internal/rbd/replicationcontrollerserver.go
Comment (BenamarMk) [inline internal/rbd/replicationcontrollerserver.go]: rbdVol is not the dummy volume.  We will be enabling mirroring on a none dummy volume if am reading this correctly.

Comment (BenamarMk) [reply internal/rbd/replicationcontrollerserver.go]: Also, enabling mirroring for the dummy volume should be done after we promote the real volume.

Comment (Madhu-1) [reply internal/rbd/replicationcontrollerserver.go]: `enableMirroringOnDummyImage` takes the rbdVol as input and creates a new dummy struct and replaces the name and calls the mirror operation on that dummy image.

Comment (Madhu-1) [reply internal/rbd/replicationcontrollerserver.go]: @BenamarMk is it the same for disabling mirroring on dummy image also? it should be done after demoting the real image?

Comment (Madhu-1) [reply internal/rbd/replicationcontrollerserver.go]: fixed for enabling. let me know your input for disable also.

Comment (BenamarMk) [reply internal/rbd/replicationcontrollerserver.go]: Disable is fine.  YOu don't need to do anything to it.  So basically the flow is like this:
1. Promote
2. enable dummy
And then
1. Disable dummy
2. Demote

File: internal/rbd/replicationcontrollerserver.go
Comment (BenamarMk) [inline internal/rbd/replicationcontrollerserver.go:540]: Shouldn't the whole block from 532 all the way to 540 be inside the block above? IOW, don't you need to put it at line 529?

Comment (Madhu-1) [reply internal/rbd/replicationcontrollerserver.go:540]: The reason for keeping it out of the loop is if the csidriver is restarted after promote and before enabling the mirroring on the dummy image on the next request the if check will be skipped as the image is already primary

File: internal/rbd/replicationcontrollerserver.go
Comment (Madhu-1) [reply internal/rbd/replicationcontrollerserver.go]: okay, then the current changes are fine. Thanks for confirming @BenamarMk

**Review by BenamarMk — APPROVED** [review]
It looks good to me @Madhu-1. Thanks for coding it so quickly.  I will now test it.

**Comment by Madhu-1** [comment]
> /retest ci/centos/upgrade-tests-cephfs

`Nov 17 15:36:49.574: FAIL: failed to create user cephcsi-cephfs-node with error etcdserver: request timed out`

File: internal/rbd/replicationcontrollerserver.go
Comment (ShyamsundarR) [inline internal/rbd/replicationcontrollerserver.go:622]: We need to ensure dummy image is primary (even if split-brain) for the enable/disable schedule to not error out. IOW, the dummy image can be in a split-brain scenario forever, but needs to be primary for these operations.

Comment (ShyamsundarR) [reply internal/rbd/replicationcontrollerserver.go:622]: Ignore @BenamarMk pointed out that we create different dummy images per cluster, so that would always be primary for the local cluster allowing enable/disable calls.

File: internal/rbd/replicationcontrollerserver.go
Comment (humblec) [inline internal/rbd/replicationcontrollerserver.go]: node/mode

File: internal/rbd/replicationcontrollerserver.go
Comment (humblec) [inline internal/rbd/replicationcontrollerserver.go]: getDummyImageName

**Comment by humblec** [comment]
One general question here @Madhu-1 , whlie the failback happen and if the dummy image exist, do we foresee any issues?

**Comment by Madhu-1** [comment]
> One general question here @Madhu-1 , whlie the failback happen and if the dummy image exist, do we foresee any issues?

Do you mean the dummy image created by the user? it's a workaround I don't think the user will create any real images with the same name for using it.

**Comment by humblec** [comment]
> > One general question here @Madhu-1 , whlie the failback happen and if the dummy image exist, do we foresee any issues?
> 
> Do you mean the dummy image created by the user? it's a workaround I don't think the user will create any real images with the same name for using it.

not that, a dummy image exist in the cluster and during failover we create another one for second cluster, Now during failback, the first cluster's dummy image exist already or not cleaned up, Thinking that,  I was asking the existence of dummy image in the first cluster can be a problem while the failback is established and images started to get promoted/demoted.

**Comment by Madhu-1** [comment]
> > > One general question here @Madhu-1 , whlie the failback happen and if the dummy image exist, do we foresee any issues?
> > 
> > 
> > Do you mean the dummy image created by the user? it's a workaround I don't think the user will create any real images with the same name for using it.
> 
> not that, a dummy image exist in the cluster and during failover we create another one for second cluster,

The dummy image is unique per cluster it's created with the fsid as the prefix.

 >Now during fallback,  the first cluster's dummy image exist already or not cleaned up, Thinking that, I was asking the existence of dummy image in the first cluster can be a problem while the failback is established and images started to get promoted/demoted.

i don't see any problem here. during fallback the dummy image mirroring will be disabled we are not going to cleanup or do anything. we operator on the dummy images which is created for that cluster not the dummy image of the other cluster.

**Comment by humblec** [comment]
> > > > One general question here @Madhu-1 , whlie the failback happen and if the dummy image exist, do we foresee any issues?
> > > 
> > > 
> > > Do you mean the dummy image created by the user? it's a workaround I don't think the user will create any real images with the same name for using it.
> > 
> > 
> > not that, a dummy image exist in the cluster and during failover we create another one for second cluster,
> 
> The dummy image is unique per cluster it's created with the fsid as the prefix.

Yeah, thats correct, what I meant above was same, may be `during failover..` was unwanted.. :) 

> 
> > Now during fallback,  the first cluster's dummy image exist already or not cleaned up, Thinking that, I was asking the existence of dummy image in the first cluster can be a problem while the failback is established and images started to get promoted/demoted.
> 
> i don't see any problem here. during fallback the dummy image mirroring will be disabled we are not going to cleanup or do anything. we operator on the dummy images which is created for that cluster not the dummy image of the other cluster.

Sure, if we dont expect any problem in this scenario, we are good.

File: internal/rbd/replicationcontrollerserver.go
Comment (humblec) [inline internal/rbd/replicationcontrollerserver.go:297]: @Madhu-1 one quick question here, if the image exist, do we need to return error? rather cant we treat it as success ?

Comment (Madhu-1) [reply internal/rbd/replicationcontrollerserver.go:297]: Yes its already exists we need to treat it as a success. Example csi deployment restarts etc.

**Comment by humblec** [comment]
lgtm.. considering the importance of this PR we have to take this in at the earliest and can do further enhancements later.. @Madhu-1 will we wait for testing result from @BenamarMk @ShyamsundarR before merge ?  

CC @ceph/ceph-csi-contributors ptal

**Comment by Madhu-1** [comment]
> lgtm.. considering the importance of this PR we have to take this in at the earliest and can do further enhancements later.. @Madhu-1 will we wait for testing result from @BenamarMk @ShyamsundarR before merge ?
> 

Yes waiting for the testing result.  but want to close up on review as changes look good to @BenamarMk  and tested locally.

> CC @ceph/ceph-csi-contributors ptal

File: internal/rbd/replicationcontrollerserver.go
Comment (ShyamsundarR) [inline internal/rbd/replicationcontrollerserver.go:545]: For the second image that we promote, this enable will become a No-OP? IOW we need to disable/enable mirroring on the dummy image for each promote of a real image, such that the real image gets scheduled.

In the current form the second enable would become a no-OP is what I think. @BenamarMk ?

Comment (Madhu-1) [reply internal/rbd/replicationcontrollerserver.go:545]: Yes, correct its a no-op. the mirroring is enabled on dummy image only once for all promote operations. The same goes for disable also

Comment (BenamarMk) [reply internal/rbd/replicationcontrollerserver.go:545]: I think I tested that and it worked. I will test it again to see if refresh happens when the dummy image is enabled for mirroring while it is already ENABLED.  But if I am wrong, then we can always do this:

Promote PV:
```
1. disable mirroring Dummy
2. promote PV
3. enable mirroring Dummy
```

Demote PV:
```
1. disable mirroring Dummy
2. demote PV
```

**Review by humblec — APPROVED** [review]

**Comment by humblec** [comment]
> > lgtm.. considering the importance of this PR we have to take this in at the earliest and can do further enhancements later.. @Madhu-1 will we wait for testing result from @BenamarMk @ShyamsundarR before merge ?
> 
> Yes waiting for the testing result. but want to close up on review as changes look good to @BenamarMk and tested locally.
> 
> > CC @ceph/ceph-csi-contributors ptal

@Madhu-1 it looks like we are still testing this, However I have placed approval to avoid any unnecessary delay once we are done with the testing to merge.

**Review by ShyamsundarR — CHANGES_REQUESTED** [review]
There are a few corner cases... hopefully this prevents the auto-merge

**Comment by humblec** [comment]
> There are a few corner cases... hopefully this prevents the auto-merge

@ShyamsundarR there is already DNM, so that should be enough to block the merge.

**Comment by Madhu-1** [comment]
closing in favor of https://github.com/ceph/ceph-csi/pull/2656

**Pull request closed without merging**
